zebra/zebra_etcd.c

```diff
@@ -0,0 +1,639 @@
+       "Remote etcd server port (1-65535)\n")
+{
+  sprintf(zetcd_glob_p->zetcd_server,"%s:%s",argv[3]->arg,argv[5]->arg);
+  zlog_debug("etcd remote connection URI:%s",zetcd_glob_p->zetcd_server);
+
+  create_grpc_channel_(zetcd_glob_p->zetcd_server);
+  zetcd_glob_p->etcd_port = atoi(argv[5]->arg);
+  return CMD_SUCCESS;
+}
+/**
+ * zetcd_build_updates
+ *
+ * Process the route entries and encode them into protobuf.
+ *
+ * */
+static void zetcd_build_updates(void)
+{
+  struct stream *s;
+  rib_dest_t *dest;
+  unsigned char *buf, *data, *buf_end;
+  size_t msg_len;
+  size_t data_len;
+  struct route_entry *re;
+
+  /*Get the buffer pointer where protobuf encoded entry to be stored.*/
+  s = zetcd_glob_p->obuf;
+
+  assert(stream_empty(s));
+  /*Check available buffer is enough.*/
+  if (STREAM_WRITEABLE(s) < FPM_MAX_MSG_LEN)
+    return;
+
+  buf = STREAM_DATA(s) + stream_get_endp(s);
+  buf_end = buf + STREAM_WRITEABLE(s);
+
+  dest = TAILQ_FIRST(&zetcd_glob_p->dest_q);
+  if (!dest)
+    return;
+  data = buf;
+
+  /*Route entry type.*/
+  re = zfpm_route_for_update(dest);
+
+  /* Protobuf encode the route entry information and return length of resulted 
+   * buffer.*/
+  data_len = zetcd_encode_route(dest, re, (char *)data,
+      buf_end - data);
+
+  assert(data_len);
+  if (data_len) {
+    msg_len = (data_len);
+    stream_forward_endp(s, msg_len);
+  }
+  else
+    return;
+  /*Remove the route entry pointer from dest_q of etcd once processed.*/
+  TAILQ_REMOVE(&zetcd_glob_p->dest_q, dest, etcd_q_entries);
+}
+/**
+ * etcd_txn
+ *
+ * Request the transaction to channel.
+ * 
+ * Return TRUE if success, else FALSE.
+ * */
+static int etcd_txn(txn_wrapper_t *txn_req)
+{
+  /*calls wrapper function.*/
+  return (txn_wrapper(zetcd_glob_p->channel->obj, zetcd_glob_p->txn_res->obj,
+      txn_req->obj));
+}
+
+/**
+ * zetcd_push_routes_cb
+ *
+ * Dequeue the route nodes from etcd dest_q, fetches the important route 
+ * information from an entry. Protobuf encode the route information and push it
+ * to etcd.
+ *
+ * @param[in] prefix pointer to prefix structure storing route IP prefix.
+ * @param[in] selected_fib best selected forwarding option from available
+ * choices for particular IP prefix.
+ * */
+static int zetcd_push_routes_cb(struct thread *thread)
+{
+  struct stream *s = NULL;
+  char *key = NULL;
+  char *value = NULL;
+  int ret = 0;
+  struct prefix *prefix = NULL;
+  rib_dest_t *dest = NULL;
+  bool recursive = false;
+  int len = 0;
+  /*unsigned int num_writes = 0;
+  bool is_put_txn_pending = false;
+  bool is_del_txn_pending = false;
+  txn_wrapper_t *txn_req = NULL;*/
+
+  zetcd_glob_p->t_push = NULL;
+  do {
+    
+    s = zetcd_glob_p->obuf;
+
+    /*Retreive route entries from queue.*/
+    dest = TAILQ_FIRST(&zetcd_glob_p->dest_q);
+    if (dest->rnode)
+    {
+      /*Get route's IP prefix and use it as key to etcd keyspace.*/
+      prefix = rib_dest_prefix(dest);
+      key = inet_ntoa(prefix->u.prefix4);
```

Comment:
Can you have a prefix for etcd keyspace here for zebra? For example, we can get a prefix string as an argument (e.g., "/zebra/india/site1" or "/zebra/us/site1"),  then use the prefix here `"/zebra/india/site1" + prefix4`. It will help to make this works compatibly with nefeli cloud.

Reply:
Yes, That is something I planned for next cycle.
I was thinking to have user optional input for `base prefix` along with `etcd URI`  from command, if not given the default base key prefix will be used. Also I need to add prefix length to the string of key, as while deleting route entry in data-path VM we require prefix and prefix length.

Reply:
Sounds great!